src/react/SessionManager.ts

```diff
@@ -0,0 +1,262 @@
+      defaultRealm?: string;
+    }
+  | {
+      auth: [{ type: 'oauth' }];
+      defaultAuth?: 'oauth';
+      defaultRealm?: string;
+    }
+  | {
+      auth: [{ type: 'token' }, { type: 'oauth' }];
+      defaultAuth?: 'token' | 'oauth';
+      defaultRealm?: string;
+    };
+
+type SessionManagerParams = {
+  astarteApiUrl: string;
+  appEngineApiUrl?: string;
+  realmManagementApiUrl?: string;
+  pairingApiUrl?: string;
+  flowApiUrl?: string;
+  enableFlowPreview: boolean;
+} & SessionManagerAuthParams;
+
+type SessionConfig = {
+  secureConnection: boolean;
+  appEngineApiUrl: URL;
+  realmManagementApiUrl: URL;
+  pairingApiUrl: URL;
+  flowApiUrl: URL;
+  enableFlowPreview: boolean;
+};
+
+type SessionCredentials = {
+  realm: string;
+  token: string;
+};
+
+type Session = {
+  config: SessionConfig;
+  credentials: SessionCredentials;
+  loginType: string | 'TokenLogin';
+};
+
+type SessionManagerEvents = {
+  sessionChange: (session: Session | null) => void;
+};
+
+function urlToSchemalessString(url: URL): string {
+  return url.toString().split(/https?:\/\//)[1];
+}
+
+function schemalessStringToURL(url: string, secureSchema: boolean): URL {
+  return new URL(secureSchema ? `https://${url}` : `http://${url}`);
+}
+
+function serializeSession(session?: Session | null): string {
+  if (!session) {
+    return 'null';
+  }
+  return JSON.stringify({
+    api_config: {
+      secure_connection: !!session.config.secureConnection,
+      realm_management_url: urlToSchemalessString(session.config.realmManagementApiUrl),
+      appengine_url: urlToSchemalessString(session.config.appEngineApiUrl),
+      pairing_url: urlToSchemalessString(session.config.pairingApiUrl),
+      flow_url: urlToSchemalessString(session.config.flowApiUrl),
+      enable_flow_preview: session.config.enableFlowPreview,
+      realm: session.credentials.realm,
+      token: session.credentials.token,
+    },
+    login_type: session.loginType,
+  });
+}
+
+function deserializeSession(serializedSession?: string | null): Session | null {
+  let session: Session | null = null;
+  try {
+    const json = JSON.parse(serializedSession || '');
+    const secureConnection = !!json.api_config.secure_connection;
+    session = {
+      config: {
+        secureConnection,
+        realmManagementApiUrl: schemalessStringToURL(
+          json.api_config.realm_management_url,
+          secureConnection,
+        ),
+        appEngineApiUrl: schemalessStringToURL(json.api_config.appengine_url, secureConnection),
+        pairingApiUrl: schemalessStringToURL(json.api_config.pairing_url, secureConnection),
+        flowApiUrl: schemalessStringToURL(json.api_config.flow_url, secureConnection),
+        enableFlowPreview: !!json.api_config.enable_flow_preview,
+      },
+      credentials: {
+        realm: json.api_config.realm,
+        token: json.api_config.token,
+      },
+      loginType: json.login_type,
+    };
+  } catch {
+    // Could not deserialize session
```

Comment:
is this ok?

Reply:
I made the behavior in case of error more obvious, specifying the meaning of the `catch` block

```javascript
try {
...
} catch {
  session = null;
}
```

Any smarter ideas are more than welcome.